sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/VarShouldBeValCheck.kt

```diff
@@ -0,0 +1,107 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtBinaryExpression
+import org.jetbrains.kotlin.psi.KtBlockExpression
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtNameReferenceExpression
+import org.jetbrains.kotlin.psi.KtOperationExpression
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtUnaryExpression
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.psi.psiUtil.findDescendantOfType
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.BindingContext.REFERENCE_TARGET
+import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.checks.AbstractCheck
+import org.sonarsource.kotlin.api.frontend.KotlinFileContext
+
+@Rule(key = "S3353")
+class VarShouldBeValCheck : AbstractCheck() {
+    override fun visitProperty(property: KtProperty, context: KotlinFileContext) {
+        if (property.isLocal && property.isVar) {
+            val bindingContext = context.bindingContext
+            val block = property.getParentOfType<KtBlockExpression>(false)!!
+            val assignments = collectAssignmentTo(block, property, bindingContext)
+
+            if (assignments.isEmpty()) {
+                context.reportIssue(property, """Replace the keyword `var` with `val`.""")
+            }
+        }
+    }
+
+    private fun collectAssignmentTo(block: PsiElement, variable: KtProperty, bindingContext: BindingContext): List<KtOperationExpression> {
+        KtDotQualifiedExpression
+        return block
+            .collectDescendantsOfType<KtOperationExpression> { expr -> isAssignment(expr) }
+            .filter { assignment ->
+                when (assignment) {
+                    is KtBinaryExpression -> isReferencingVar(variable, getReference(assignment), bindingContext) ?: false
+                    is KtUnaryExpression -> isReferencingVar(variable, getReference(assignment), bindingContext)
+                    // I don't know how to cover it
+                    else -> false
+                }
+            }
```

Comment:
What do you think about explicitly collecting descendants of types `KtUnaryExpression` and `KtBinaryExpression`? This way it should be possible to cover with tests all possible cases :)
```suggestion
        return listOf(
            block
                .collectDescendantsOfType<KtUnaryExpression> { expr -> isAssignment(expr) }
                .filter { assignment -> isAssignment(assignment) && isReferencingVar(variable, getReference(assignment), bindingContext) },
            block
                .collectDescendantsOfType<KtBinaryExpression> { expr -> isAssignment(expr) }
                .filter { assignment -> isAssignment(assignment) && isReferencingVar(variable, getReference(assignment), bindingContext) }
        ).flatten()
```

Please, feel free to rework this code to reduce duplications 🤞

Reply:
This could even be simplified to return a Boolean and the two `filter` calls replaced by calls to `none` with an && operation. Then the `visitProperty` doesn't need to check for emptiness on the returned collection

Reply:
:+1: